src/Stratis.Bitcoin.Features.PoA/Voting/IdleFederationMembersKicker.cs

```diff
@@ -0,0 +1,144 @@
+using Stratis.Bitcoin.Signals;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.PoA.Voting
+{
+    /// <summary>
+    /// Automatically schedules addition of voting data that votes for kicking federation member that
+    /// didn't produce a block in <see cref="PoAConsensusOptions.FederationMemberMaxIdleTimeSeconds"/>.
+    /// </summary>
+    public class IdleFederationMembersKicker : IDisposable
+    {
+        private readonly ISignals signals;
+
+        private readonly IKeyValueRepository keyValueRepository;
+
+        private readonly IConsensusManager consensusManager;
+
+        private readonly Network network;
+
+        private readonly FederationManager federationManager;
+
+        private readonly SlotsManager slotsManager;
+
+        private readonly VotingManager votingManager;
+
+        private readonly ILogger logger;
+
+        private readonly uint federationMemberMaxIdleTimeSeconds;
+
+        private SubscriptionToken blockConnectedToken, fedMemberAddedToken, fedMemberKickedToken;
+
+        /// <remarks>Active time is updated when member is added or produced a new block.</remarks>
+        private Dictionary<PubKey, uint> fedMembersByLastActiveTime;
+
+        private const string fedMembersByLastActiveTimeKey = "fedMembersByLastActiveTime";
+
+        public IdleFederationMembersKicker(ISignals signals, Network network, IKeyValueRepository keyValueRepository, IConsensusManager consensusManager,
+            FederationManager federationManager, SlotsManager slotsManager, VotingManager votingManager, ILoggerFactory loggerFactory)
+        {
+            this.signals = signals;
+            this.network = network;
+            this.keyValueRepository = keyValueRepository;
+            this.consensusManager = consensusManager;
+            this.federationManager = federationManager;
+            this.slotsManager = slotsManager;
+            this.votingManager = votingManager;
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.federationMemberMaxIdleTimeSeconds = ((PoAConsensusOptions)network.Consensus.Options).FederationMemberMaxIdleTimeSeconds;
+        }
+
+        public void Initialize()
+        {
+            this.blockConnectedToken = this.signals.Subscribe<BlockConnected>(this.OnBlockConnected);
+            this.fedMemberAddedToken = this.signals.Subscribe<FedMemberAdded>(this.OnFedMemberAdded);
+            this.fedMemberKickedToken = this.signals.Subscribe<FedMemberKicked>(this.OnFedMemberKicked);
+
+            this.fedMembersByLastActiveTime = this.keyValueRepository.LoadValueJson<Dictionary<PubKey, uint>>(fedMembersByLastActiveTimeKey);
+
+            if (this.fedMembersByLastActiveTime == null)
+            {
+                this.logger.LogDebug("No saved data found. Initializing federation data with genesis timestamps.");
+
+                this.fedMembersByLastActiveTime = new Dictionary<PubKey, uint>();
```

Comment:
Why not init in the constructor?

Reply:
To have an exception if someone attempts to use the component while this feature is disabled (initialize is not called if its disabled).